File: snippets/auth/user-roles.py
"""
SNIPPET: User Roles & Whitelisting
CATEGORY: Authentication & Access Control
LANGUAGE: Python (FastAPI)
STATUS: Ready

DESCRIPTION:
    Role-based access control with admin approval workflow.
    Auto-creates users on first login from OIDC headers.
    Supports user/expert/admin roles and pending/approved/denied statuses.

DEPENDENCIES:
    pip install asyncpg

ENVIRONMENT VARIABLES:
    SECRET_3 or INITIAL_ADMIN_EMAILS — comma-separated bootstrap admin emails

SQL SCHEMA:
    Uses the same schema as auth/user-roles-schema.sql

USAGE:
    from user_roles import create_session, require_approved, require_role, auth_router

    # Auto-create user on login
    @app.middleware("http")
    async def auth_middleware(request, call_next):
        await create_session(request)
        return await call_next(request)

    # Protect routes
    @app.get("/dashboard", dependencies=[Depends(require_approved)])
    async def dashboard(): ...

    @app.get("/admin", dependencies=[Depends(require_role("admin"))])
    async def admin_panel(): ...

    # Mount auth routes
    app.include_router(auth_router, prefix="/auth")

RELATED:
    - snippets/auth/user-roles-schema.sql (database schema)
    - snippets/auth/user-identity.py (basic OIDC identity extraction)
"""

import os
from fastapi import APIRouter, Depends, HTTPException, Request
import asyncpg

# ---------------------------------------------------------------------------
# Config
# ---------------------------------------------------------------------------
ADMIN_EMAILS = [
    e.strip().lower()
    for e in (os.getenv("INITIAL_ADMIN_EMAILS") or os.getenv("SECRET_3") or "").split(",")
    if e.strip()
]

_pool = None

async def get_pool():
    global _pool
    if _pool is None:
        _pool = await asyncpg.create_pool(
            host=os.getenv("DATABASE_HOST", "db"),
            port=int(os.getenv("DATABASE_PORT", "5432")),
            database=os.getenv("DATABASE_NAME", "myapp"),
            user=os.getenv("DATABASE_USER", "postgres"),
            password=os.getenv("DATABASE_PASSWORD", "postgres"),
        )
    return _pool

# ---------------------------------------------------------------------------
# Session management
# ---------------------------------------------------------------------------

async def create_session(request: Request) -> dict | None:
    """Upsert user from OIDC headers. Auto-approves admin emails."""
    oidc_id = request.headers.get("x-amzn-oidc-identity")
    email = request.headers.get("x-amzn-oidc-email", "").lower()
    name = request.headers.get("x-amzn-oidc-name", "")

    if not oidc_id:
        # Local dev fallback
        if os.getenv("NODE_ENV") != "production":
            request.state.user = {"id": 0, "email": "dev@local", "name": "Dev User", "role": "admin", "account_status": "approved"}
            return request.state.user
        return None

    pool = await get_pool()

    # Check if user exists
    row = await pool.fetchrow("SELECT * FROM users WHERE oidc_user_id = $1", oidc_id)

    if row:
        request.state.user = dict(row)
        return dict(row)

    # New user — auto-approve admins
    is_admin = email in ADMIN_EMAILS
    role = "admin" if is_admin else "user"
    status = "approved" if is_admin else "pending"

    row = await pool.fetchrow(
        """INSERT INTO users (oidc_user_id, email, name, role, account_status)
           VALUES ($1, $2, $3, $4, $5)
           RETURNING *""",
        oidc_id, email, name, role, status,
    )
    request.state.user = dict(row)
    return dict(row)

# ---------------------------------------------------------------------------
# Dependencies
# ---------------------------------------------------------------------------

async def require_approved(request: Request):
    """FastAPI dependency: block pending/denied users."""
    user = getattr(request.state, "user", None)
    if not user:
        raise HTTPException(401, "Not authenticated")
    if user.get("account_status") != "approved":
        raise HTTPException(403, f"Account is {user.get('account_status', 'unknown')}")
    return user

def require_role(*roles: str):
    """FastAPI dependency factory: check user role."""
    async def check_role(request: Request):
        user = getattr(request.state, "user", None)
        if not user:
            raise HTTPException(401, "Not authenticated")
        if user.get("account_status") != "approved":
            raise HTTPException(403, f"Account is {user.get('account_status', 'unknown')}")
        if user.get("role") not in roles:
            raise HTTPException(403, f"Requires role: {', '.join(roles)}")
        return user
    return Depends(check_role)

# ---------------------------------------------------------------------------
# Admin routes
# ---------------------------------------------------------------------------
auth_router = APIRouter()

@auth_router.get("/users")
async def list_users(_=Depends(require_role("admin"))):
    pool = await get_pool()
    rows = await pool.fetch("SELECT id, email, name, role, account_status, created_at FROM users ORDER BY created_at DESC")
    return [dict(r) for r in rows]

@auth_router.put("/users/{user_id}/role")
async def update_role(user_id: int, body: dict, _=Depends(require_role("admin"))):
    role = body.get("role")
    if role not in ("user", "expert", "admin"):
        raise HTTPException(400, "Invalid role")
    pool = await get_pool()
    await pool.execute("UPDATE users SET role = $1 WHERE id = $2", role, user_id)
    return {"success": True}

@auth_router.put("/users/{user_id}/access")
async def update_access(user_id: int, body: dict, _=Depends(require_role("admin"))):
    status = body.get("status")
    if status not in ("approved", "denied", "pending"):
        raise HTTPException(400, "Invalid status")
    pool = await get_pool()
    await pool.execute("UPDATE users SET account_status = $1 WHERE id = $2", status, user_id)
    return {"success": True}

File: snippets/auth/user-roles.js
/**
 * SNIPPET: User Roles & Access Control (Express + PostgreSQL)
 * CATEGORY: Authentication
 * LANGUAGE: JavaScript (Express)
 * STATUS: Ready
 *
 * DESCRIPTION:
 *   Role-based access control with admin approval workflow.
 *   Provides session creation (upsert on login), middleware for
 *   role/status checks, audit logging, and an Express router
 *   for user management endpoints.
 *
 *   Roles: user, expert, admin
 *   Account statuses: pending, approved, denied
 *
 *   Bootstrap admins are auto-approved on first login via the
 *   INITIAL_ADMIN_EMAILS (or SECRET_3) environment variable.
 *
 * DEPENDENCIES:
 *   npm install pg
 *
 * ENV:
 *   SECRET_3 or INITIAL_ADMIN_EMAILS — comma-separated list of
 *   bootstrap admin email addresses (auto-approved + admin role)
 *
 * USAGE:
 *   const { createUserRouter, requireApproved, requireRole } = require('./user-roles');
 *
 *   // Attach after userIdentityMiddleware (see user-identity.js)
 *   app.use('/auth', createUserRouter(pool));
 *
 *   // Protect routes
 *   app.use('/api', requireApproved(pool));
 *   app.get('/api/admin/stats', requireRole(pool, 'admin'), handler);
 *
 * RELATED:
 *   - snippets/auth/user-roles-schema.sql (database schema)
 *   - snippets/auth/UserManagement.jsx (admin UI)
 *   - snippets/auth/user-identity.js (OIDC identity extraction)
 */

const express = require('express');

// ============================================
// Constants
// ============================================

const VALID_ROLES = ['user', 'expert', 'admin'];
const VALID_STATUSES = ['pending', 'approved', 'denied'];

/**
 * Comma-separated list of bootstrap admin emails.
 * These users are auto-approved with admin role on first login.
 */
const ADMIN_EMAILS = (process.env.SECRET_3 || process.env.INITIAL_ADMIN_EMAILS || '')
  .split(',')
  .map(e => e.trim().toLowerCase())
  .filter(Boolean);

// ============================================
// Core Functions
// ============================================

/**
 * Create or update a user session on login.
 * If the user's email is in ADMIN_EMAILS, they are auto-approved as admin.
 * Uses INSERT...ON CONFLICT (upsert) with RETURNING for atomicity.
 *
 * @param {import('pg').Pool} pool - PostgreSQL connection pool
 * @param {Object} req - Express request (must have req.user from OIDC middleware)
 * @returns {Promise<Object>} The user record from the database
 */
async function createSession(pool, req) {
  const { userId, email, name } = req.user;
  const isBootstrapAdmin = email && ADMIN_EMAILS.includes(email.toLowerCase());

  const role = isBootstrapAdmin ? 'admin' : 'user';
  const status = isBootstrapAdmin ? 'approved' : 'pending';

  const result = await pool.query(
    `INSERT INTO users (oidc_user_id, email, name, role, account_status, last_login)
     VALUES ($1, $2, $3, $4, $5, NOW())
     ON CONFLICT (oidc_user_id) DO UPDATE SET
       email = COALESCE(EXCLUDED.email, users.email),
       name = COALESCE(EXCLUDED.name, users.name),
       last_login = NOW()
     RETURNING *`,
    [userId, email, name, role, status]
  );

  return result.rows[0];
}

/**
 * Express middleware that requires the user's account to be approved.
 * Returns 403 if the account is pending or denied.
 *
 * @param {import('pg').Pool} pool - PostgreSQL connection pool
 * @returns {Function} Express middleware
 */
function requireApproved(pool) {
  return async (req, res, next) => {
    try {
      const result = await pool.query(
        'SELECT account_status FROM users WHERE oidc_user_id = $1',
        [req.user.userId]
      );

      if (!result.rows.length) {
        return res.status(403).json({ error: 'User not found. Please log in first.' });
      }

      if (result.rows[0].account_status !== 'approved') {
        return res.status(403).json({
          error: 'Account not approved',
          status: result.rows[0].account_status
        });
      }

      next();
    } catch (err) {
      console.error('requireApproved error:', err);
      res.status(500).json({ error: 'Internal server error' });
    }
  };
}

/**
 * Express middleware factory that checks the user's role against allowed roles.
 * Must be used after requireApproved() to ensure the user exists.
 *
 * @param {import('pg').Pool} pool - PostgreSQL connection pool
 * @param {...string} roles - Allowed roles (e.g., 'admin', 'expert')
 * @returns {Function} Express middleware
 */
function requireRole(pool, ...roles) {
  return async (req, res, next) => {
    try {
      const result = await pool.query(
        'SELECT role FROM users WHERE oidc_user_id = $1',
        [req.user.userId]
      );

      if (!result.rows.length) {
        return res.status(403).json({ error: 'User not found' });
      }

      if (!roles.includes(result.rows[0].role)) {
        return res.status(403).json({ error: 'Insufficient permissions' });
      }

      next();
    } catch (err) {
      console.error('requireRole error:', err);
      res.status(500).json({ error: 'Internal server error' });
    }
  };
}

/**
 * Insert an entry into the audit log.
 *
 * @param {import('pg').Pool} pool - PostgreSQL connection pool
 * @param {Object} entry - Audit log entry
 * @param {number} entry.userId - ID of the user performing the action
 * @param {string} entry.action - Action name (e.g., 'approve_user', 'change_role')
 * @param {string} [entry.entityType] - Type of entity acted upon (e.g., 'user')
 * @param {string} [entry.entityId] - ID of the entity acted upon
 * @param {Object} [entry.details] - Additional details (stored as JSONB)
 */
async function logAudit(pool, { userId, action, entityType, entityId, details }) {
  await pool.query(
    `INSERT INTO audit_log (user_id, action, entity_type, entity_id, details)
     VALUES ($1, $2, $3, $4, $5)`,
    [userId, action, entityType || null, entityId || null, JSON.stringify(details || {})]
  );
}

// ============================================
// Express Router
// ============================================

/**
 * Create an Express router for user management endpoints.
 *
 * @param {import('pg').Pool} pool - PostgreSQL connection pool
 * @returns {import('express').Router}
 */
function createUserRouter(pool) {
  const router = express.Router();

  // GET / — return the current user's identity (from OIDC middleware)
  router.get('/', (req, res) => {
    res.json(req.user);
  });

  // POST /session — upsert user on login, return DB record
  router.post('/session', async (req, res) => {
    try {
      const dbUser = await createSession(pool, req);
      res.json(dbUser);
    } catch (err) {
      console.error('POST /session error:', err);
      res.status(500).json({ error: 'Failed to create session' });
    }
  });

  // GET /users — list all users (admin only), optional ?account_status filter
  router.get('/users', requireRole(pool, 'admin'), async (req, res) => {
    try {
      const { account_status } = req.query;
      let query = 'SELECT * FROM users ORDER BY created_at DESC';
      const params = [];

      if (account_status && VALID_STATUSES.includes(account_status)) {
        query = 'SELECT * FROM users WHERE account_status = $1 ORDER BY created_at DESC';
        params.push(account_status);
      }

      const result = await pool.query(query, params);
      res.json(result.rows);
    } catch (err) {
      console.error('GET /users error:', err);
      res.status(500).json({ error: 'Failed to fetch users' });
    }
  });

  // PUT /users/:id/role — update a user's role (admin only)
  router.put('/users/:id/role', requireRole(pool, 'admin'), async (req, res) => {
    try {
      const { role } = req.body;
      if (!VALID_ROLES.includes(role)) {
        return res.status(400).json({ error: `Invalid role. Must be one of: ${VALID_ROLES.join(', ')}` });
      }

      const result = await pool.query(
        'UPDATE users SET role = $1 WHERE id = $2 RETURNING *',
        [role, req.params.id]
      );

      if (!result.rows.length) {
        return res.status(404).json({ error: 'User not found' });
      }

      // Get admin's DB user for audit log
      const admin = await pool.query('SELECT id FROM users WHERE oidc_user_id = $1', [req.user.userId]);
      if (admin.rows.length) {
        await logAudit(pool, {
          userId: admin.rows[0].id,
          action: 'change_role',
          entityType: 'user',
          entityId: req.params.id,
          details: { newRole: role }
        });
      }

      res.json(result.rows[0]);
    } catch (err) {
      console.error('PUT /users/:id/role error:', err);
      res.status(500).json({ error: 'Failed to update role' });
    }
  });

  // PUT /users/:id/access — approve or deny a user (admin only)
  router.put('/users/:id/access', requireRole(pool, 'admin'), async (req, res) => {
    try {
      const { account_status } = req.body;
      if (!['approved', 'denied'].includes(account_status)) {
        return res.status(400).json({ error: 'Invalid status. Must be approved or denied.' });
      }

      const result = await pool.query(
        'UPDATE users SET account_status = $1 WHERE id = $2 RETURNING *',
        [account_status, req.params.id]
      );

      if (!result.rows.length) {
        return res.status(404).json({ error: 'User not found' });
      }

      const admin = await pool.query('SELECT id FROM users WHERE oidc_user_id = $1', [req.user.userId]);
      if (admin.rows.length) {
        await logAudit(pool, {
          userId: admin.rows[0].id,
          action: account_status === 'approved' ? 'approve_user' : 'deny_user',
          entityType: 'user',
          entityId: req.params.id,
          details: { newStatus: account_status }
        });
      }

      res.json(result.rows[0]);
    } catch (err) {
      console.error('PUT /users/:id/access error:', err);
      res.status(500).json({ error: 'Failed to update access' });
    }
  });

  return router;
}

module.exports = {
  ADMIN_EMAILS,
  createSession,
  requireApproved,
  requireRole,
  logAudit,
  createUserRouter
};

// ============================================
// Frontend Pattern: Role-Based Navigation
// ============================================
//
// Filter sidebar items by user role so restricted pages
// are hidden from users who don't have access:
//
//   const navItems = [
//     { path: '/', label: 'Home', icon: 'H' },
//     { path: '/admin', label: 'Admin', icon: 'A', roles: ['admin'] },
//   ];
//   const visible = navItems.filter(i => !i.roles || i.roles.includes(userRole));
//
